hs_file_types/models/reftimeseries.py

```diff
@@ -0,0 +1,922 @@
+        return ['WOF', "WPS", "DirectFile"]
+
+    @classmethod
+    def get_allowed_return_types(cls):
+        """returns a list of returnType controlled vocabulary terms"""
+        return ['WaterML 1.1', 'WaterML 2.0', 'TimeseriesML']
+
+    @classmethod
+    def get_allowed_service_types(cls):
+        """returns a list of serviceType controlled vocabulary terms"""
+        return ['SOAP', 'REST']
+
+    @classmethod
+    def create(cls):
+        # this custom method MUST be used to create an instance of this class
+        rf_ts_metadata = RefTimeseriesFileMetaData.objects.create(json_file_content="No data")
+        return cls.objects.create(metadata=rf_ts_metadata)
+
+    @classmethod
+    def set_file_type(cls, resource, file_id, user):
+        """
+            Sets a json resource file to RefTimeseriesFile type
+            :param resource: an instance of resource type CompositeResource
+            :param file_id: id of the resource file to be set as RefTimeSeriesFile type
+            :param user: user who is setting the file type
+            :return:
+            """
+
+        log = logging.getLogger()
+
+        # get the the selected resource file object
+        res_file = utils.get_resource_file_by_id(resource, file_id)
+
+        if res_file is None:
+            raise ValidationError("File not found.")
+
+        if res_file.extension != '.refts':
+            raise ValidationError("Not a Ref Time Series file.")
+
+        files_to_add_to_resource = []
+        if res_file.has_generic_logical_file:
+            try:
+                json_file_content = _validate_json_file(res_file)
+            except Exception as ex:
+                raise ValidationError(ex.message)
+
+            # get the file from irods to temp dir
+            temp_file = utils.get_file_from_irods(res_file)
+            temp_dir = os.path.dirname(temp_file)
+            files_to_add_to_resource.append(temp_file)
+            file_folder = res_file.file_folder
+            with transaction.atomic():
+                # first delete the json file that we retrieved from irods
+                # for setting it to reftimeseries file type
+                delete_resource_file(resource.short_id, res_file.id, user)
+
+                # create a reftiemseries logical file object to be associated with
+                # resource files
+                logical_file = cls.create()
+
+                logical_file.metadata.json_file_content = json_file_content
+                logical_file.metadata.save()
+
+                try:
+                    # add the json file back to the resource
+                    uploaded_file = UploadedFile(file=open(temp_file, 'rb'),
+                                                 name=os.path.basename(temp_file))
+
+                    new_res_file = utils.add_file_to_resource(
+                        resource, uploaded_file, folder=file_folder
+                    )
```

Comment:
@pkdash I feel this workflow to get file from irods to local temp, delete file from irods, upload local temp file, then add this file to resource probably can be simplified for performance reasons. Since this workflow applies to all logical file type setting for composite resources, I assume changing/improving this would take a lot of work, so I am not asking for making this change, but just a food for thought, since doing this workflow for large files such as a gigabyte size raster or netcdf file is costly. If doing this workflow is only for metadata extraction and leveraging existing code, I think with some careful design, this can be achieved without transferring files back and forth as done currently. What do you think?